ios-sockets/SwiftWebSockets/SwiftWebSockets/Networking/NWWebSocket.swift

```diff
@@ -182,53 +207,50 @@ open class NWWebSocket: WebSocketConnection {
+                         isComplete: true,
+                         completion: .contentProcessed({ [weak self] error in
                             guard let self = self else {
                                 return
                             }
 
-                            if let error = error {
-                                self.connectionDidFail(error: error)
+                            // If a connection closure was sent, inform delegate on completion
+                            if let socketMetadata = context.protocolMetadata.first as? NWProtocolWebSocket.Metadata,
+                                socketMetadata.opcode == .close {
+                                self.delegate?.webSocketDidDisconnect(connection: self,
+                                                                      closeCode: socketMetadata.closeCode,
+                                                                      reason: data)
                             }
-                        }))
-    }
 
-    private func connectionDidFail(error: NWError) {
-        delegate?.webSocketDidReceiveError(connection: self, error: error)
-        stop(error: error)
+                            if let error = error {
+                                self.delegate?.webSocketDidReceiveError(connection: self, error: error)
+                            }
+                         }))
     }
 
-    private func stateDidChange(to state: NWConnection.State) {
-        switch state {
-        case .ready:
-            delegate?.webSocketDidConnect(connection: self)
-        case .waiting(let error), .failed(let error):
-            connectionDidFail(error: error)
-        case .setup:
-            //
-            break
-        case .preparing:
-            //
-            break
-        case .cancelled:
-            //
-            break
-        @unknown default:
-            fatalError()
+    private func stopConnection(error: NWError?) {
+        if let error = error, shouldReportNWError(error) {
+            delegate?.webSocketDidReceiveError(connection: self, error: error)
         }
+        pingTimer?.invalidate()
+        connection = nil
     }
 
-    private func connectionDidEnd(closeCode: NWProtocolWebSocket.CloseCode, reason: Data?) {
-        delegate?.webSocketDidDisconnect(connection: self, closeCode: closeCode, reason: reason)
-        stop(error: nil)
-    }
-
-    private func stop(error: Error?) {
-        connection.stateUpdateHandler = nil
-        connection.cancel()
+    /// Determine if an Network error should be reported.
+    ///
+    /// POSIX errors of either `ENOTCONN` ("Socket is not connected") or
+    /// `ECANCELED` ("Operation canceled") should not be reported if the disconnection was intentional.
+    /// All other errors should be reported.
+    /// - Parameter error: An `NWError` to inspect.
+    /// - Returns: `true` if the error should be reported.
+    private func shouldReportNWError(_ error: NWError) -> Bool {
```

Comment:
@danielrbrowne were you seeing errors being thrown during termination?

Reply:
@MichaelNeas Yes in my testing, these two error cases are reported during disconnection events (even if the disconnection is intentional).

`.ECANCELED` is thrown if `disconnect(closeCode:)` is called with a `.protocolCode(.normalClosure)` close code, and `.ENOTCONN` is thrown if ``disconnect(closeCode:)` is called with any other close code.

For my purposes, I don't want these errors reported (i.e. I don't consider them errors), but it may be desirable to make the logic configurable rather than just checking `intentionalDisconnect == true`?

Reply:
Interesting, I think a configurable solution would be good but the `intentionalDisconnect` flag sounds like it would benefit most users

Reply:
Agreed, this could be a first issue on a new repo :)